compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -819,15 +815,33 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
     }
 
     fn bitcast(&mut self, val: &'ll Value, dest_ty: &'ll Type) -> &'ll Value {
-        unsafe { llvm::LLVMBuildBitCast(self.llbuilder, val, dest_ty, UNNAMED) }
+        unsafe {
+            if llvm::LLVMTypeOf(val) != dest_ty {
+                llvm::LLVMBuildBitCast(self.llbuilder, val, dest_ty, UNNAMED)
+            } else {
+                val
+            }
+        }
     }
 
     fn intcast(&mut self, val: &'ll Value, dest_ty: &'ll Type, is_signed: bool) -> &'ll Value {
-        unsafe { llvm::LLVMRustBuildIntCast(self.llbuilder, val, dest_ty, is_signed) }
+        unsafe {
+            if llvm::LLVMTypeOf(val) != dest_ty {
+                llvm::LLVMRustBuildIntCast(self.llbuilder, val, dest_ty, is_signed)
+            } else {
+                val
+            }
+        }
     }
 
     fn pointercast(&mut self, val: &'ll Value, dest_ty: &'ll Type) -> &'ll Value {
-        unsafe { llvm::LLVMBuildPointerCast(self.llbuilder, val, dest_ty, UNNAMED) }
+        unsafe {
+            if llvm::LLVMTypeOf(val) != dest_ty {
+                llvm::LLVMBuildPointerCast(self.llbuilder, val, dest_ty, UNNAMED)
+            } else {
+                val
+            }
```

Comment:
Doesn't builder fold this at construction already https://github.com/rust-lang/llvm-project/blob/3dfd4d93fa013e1c0578d3ceac5c8f4ebba4b6ec/llvm/include/llvm/IR/IRBuilder.h#L1998-L2005?